guide/guide-qip.rst

```diff
@@ -15,28 +15,55 @@ The Quantum Information Processing (QIP) module aims at providing basic tools fo
+To get the matrix representation of each gate, we can call the class method :meth:`qutip.qip.QubitCircuit.propagators()`. Carrying out the matrices product, one gets the matrix representation of the whole evolution. This process is demonstrated in the following example.
+
+We can also carry out measurements on individual qubits (both in the middle and at the end of the circuit).
+Each measurement is saved as a class object :class:`qutip.qip.Measurement` with parameters such as `target`,
+the target qubit on which the measurement will be carried out and `classical_store`,
+the index of the classical register which stores the measurement result.
+
+Finally, once we have constructed the circuit, we can use the
+`qutip.qip.QubitCircuit.run()` function to carry out one run of the circuit from start to finish which will return the final state as well as the probability of that state being outputs.
+Moreover, `qutip.qip.QubitCircuit.run_statistics()` function can return all the possible output states along with the respective
+probabilities.
 
 .. code-block:: python
-   
-   >>> from qutip.qip.circuit import QubitCircuit, Gate
-   >>> from qutip.qip.operations import gate_sequence_product
-   >>> qc = QubitCircuit(N=2)
-   >>> swap_gate = Gate(name="SWAP", targets=[0, 1])
-   >>> qc.add_gate(swap_gate)
-   >>> qc.add_gate("CNOT", controls=0, targets=1)
-   >>> qc.add_gate(swap_gate)
-   >>> print(qc.gates)
-   [Gate(SWAP, targets=[0, 1], controls=None), Gate(CNOT, targets=[1],
-   controls=[0]), Gate(SWAP, targets=[0, 1], controls=None)]
-   >>> U_list = qc.propagators()
-   >>> print(gate_sequence_product(U_list))
-   Quantum object: dims = [[2, 2], [2, 2]], shape = (4, 4), type = oper, isherm = True
-   Qobj data =
-   [[1. 0. 0. 0.]
-   [0. 0. 0. 1.]
-   [0. 0. 1. 0.]
-   [0. 1. 0. 0.]]
+
+  >>> from qutip.qip.circuit import QubitCircuit, Gate
+  >>> from qutip import tensor, basis
+  >>> qc = QubitCircuit(N=2, num_cbits=1)
+  >>> swap_gate = Gate(name="SWAP", targets=[0, 1])
+  >>> qc.add_gate(swap_gate)
+  >>> qc.add_measurement("M0", targets=[1], classical_store=0)
+  >>> qc.add_gate("CNOT", controls=0, targets=1)
+  >>> qc.add_gate("X", targets=0, classical_controls=[0])
+  >>> qc.add_gate(swap_gate)
+  >>> print(qc.gates)
+  [Gate(SWAP, targets=[0, 1], controls=None, classical controls=None), Gate(CNOT, targets=[1], controls=[0], classical controls=None), Gate(X, targets=[0], controls=None, classical controls=[0]), Gate(SWAP, targets=[0, 1], controls=None, classical controls=None)]
+  >>> U_list = qc.propagators()
+  >>> print(print(gate_sequence_product(U_list)))
+  Quantum object: dims = [[2, 2], [2, 2]], shape = (4, 4), type = oper, isherm = True
+  Qobj data =
+  [[1. 0. 0. 0.]
+  [0. 0. 0. 1.]
+  [0. 0. 1. 0.]
+  [0. 1. 0. 0.]]
+  >>> probability, final_state = qc.run(state=tensor(basis(2, 0), basis(2, 1)))
```

Comment:
I will try to switch to testdoc!

Reply:
My issue right now is I cannot build the documentation locally because of the ipython directive error that has been popping up since before. @nathanshammah, was there any solution to this that was arrived at finally?